File: java/java_demo/src/data_structures/hashtables/BuiltInHashTables.java
package data_structures.hashtables;

import java.util.*;

import java.util.concurrent.ConcurrentHashMap;

public class BuiltInHashTables {
    private HashMap<String, String> hashMapTable;
    private Hashtable<String, String> hashTable;
    private LinkedHashMap<String, String> linkedHashMapTable;
    private ConcurrentHashMap<String, String> concurrentHashTable;

    public BuiltInHashTables() {
        // Initialize different hash table implementations
        hashMapTable = new HashMap<>();
        hashTable = new Hashtable<>();
        linkedHashMapTable = new LinkedHashMap<>();
        concurrentHashTable = new ConcurrentHashMap<>();
    }

    // HashMap methods
    public void hashMapInsert(String key, String value) {
        hashMapTable.put(key, value);
    }

    public String hashMapGet(String key) {
        return hashMapTable.getOrDefault(key, null);
    }

    // Hashtable methods
    public void hashTableInsert(String key, String value) {
        hashTable.put(key, value);
    }

    public String hashTableGet(String key) {
        return hashTable.getOrDefault(key, null);
    }

    // LinkedHashMap methods (maintains insertion order)
    public void linkedHashMapInsert(String key, String value) {
        linkedHashMapTable.put(key, value);
    }

    public String linkedHashMapGet(String key) {
        return linkedHashMapTable.getOrDefault(key, null);
    }

    // ConcurrentHashMap methods (thread-safe)
    public void concurrentHashMapInsert(String key, String value) {
        concurrentHashTable.put(key, value);
    }

    public String concurrentHashMapGet(String key) {
        return concurrentHashTable.getOrDefault(key, null);
    }

    // Display all tables for debugging purposes
    public void displayTables() {
        System.out.println("HashMap Table: " + hashMapTable);
        System.out.println("Hashtable Table: " + hashTable);
        System.out.println("LinkedHashMap Table: " + linkedHashMapTable);
        System.out.println("ConcurrentHashMap Table: " + concurrentHashTable);
    }

    public static void demo() {

        System.out.println("==============================");
        System.out.println("Demo of built in hashtable implementations");
        System.out.println("==============================");
        BuiltInHashTables hashTables = new BuiltInHashTables();

        // Using HashMap
        hashTables.hashMapInsert("name", "Alice");
        System.out.println("HashMap Get: " + hashTables.hashMapGet("name")); // Output: Alice

        // Using Hashtable
        hashTables.hashTableInsert("age", "30");
        System.out.println("Hashtable Get: " + hashTables.hashTableGet("age")); // Output: 30

        // Using LinkedHashMap
        hashTables.linkedHashMapInsert("city", "New York");
        System.out.println("LinkedHashMap Get: " + hashTables.linkedHashMapGet("city")); // Output: New York

        // Using ConcurrentHashMap
        hashTables.concurrentHashMapInsert("country", "USA");
        System.out.println("ConcurrentHashMap Get: " + hashTables.concurrentHashMapGet("country")); // Output: USA

        hashTables.displayTables();
    }
}

File: python/python_demo/data_structures/hashtables/builtin_hashtables.py
from collections import OrderedDict, defaultdict

class BuiltInHashTables:
    """A class demonstrating different built-in hash table implementations in Python."""

    def __init__(self):
        # Initialize different types of hash tables
        self.dict_table = {}
        self.ordered_dict_table = OrderedDict()
        self.default_dict_table = defaultdict(lambda: "Default Value")

    def dict_insert(self, key, value):
        """Insert a key-value pair into a Python dictionary."""
        self.dict_table[key] = value

    def dict_get(self, key):
        """Retrieve a value from a Python dictionary."""
        return self.dict_table.get(key, None)

    def ordered_dict_insert(self, key, value):
        """Insert a key-value pair into an OrderedDict (maintains insertion order)."""
        self.ordered_dict_table[key] = value

    def ordered_dict_get(self, key):
        """Retrieve a value from an OrderedDict."""
        return self.ordered_dict_table.get(key, None)

    def default_dict_insert(self, key, value):
        """Insert a key-value pair into a defaultdict."""
        self.default_dict_table[key] = value

    def default_dict_get(self, key):
        """Retrieve a value from a defaultdict (returns default value if key is missing)."""
        return self.default_dict_table[key]  # Will return default if key doesn't exist

    def display_tables(self):
        """Print all tables for debugging purposes."""
        print("Dict Table:", self.dict_table)
        print("OrderedDict Table:", self.ordered_dict_table)
        print("DefaultDict Table:", self.default_dict_table)


# Example Usage
def demo_builtin_hashtables():
    print("============================")
    print("Demo of built in hash tables")
    print("============================")
    hash_tables = BuiltInHashTables()

    # Using dict
    hash_tables.dict_insert("name", "Alice")
    print("Dict Get:", hash_tables.dict_get("name"))  # Output: Alice

    # Using OrderedDict
    hash_tables.ordered_dict_insert("age", 30)
    print("OrderedDict Get:", hash_tables.ordered_dict_get("age"))  # Output: 30

    # Using defaultdict
    hash_tables.default_dict_insert("city", "New York")
    print("DefaultDict Get (existing key):", hash_tables.default_dict_get("city"))  # Output: New York
    print("DefaultDict Get (missing key):", hash_tables.default_dict_get("country"))  # Output: Default Value

    hash_tables.display_tables()
